nixos/tests/systemd.nix

```diff
@@ -49,6 +49,40 @@ import ./make-test-python.nix ({ pkgs, ... }: {
       '';
     };
 
+    systemd.services.test-options-merging = {
+      description = "Test Options Merging";
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        set -eux
+        test -d /var/lib/test-options-merging/foo
+        test -d /var/lib/test-options-merging/bar
+      '';
+      serviceConfig = lib.mkMerge [
+        {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        }
+        # Merging must lift non-list definitions to a list
+        # if at least one of them is a list.
+        { StateDirectory = "test-options-merging/foo"; }
+        { StateDirectory = ["test-options-merging/bar"]; }
+      ];
+    };
+
+    systemd.services.test-options-merging-conflict = {
+      enable =
+        assert !(builtins.tryEval (config.systemd.services.test-options-merging-conflict.serviceConfig.StateDirectory)).success;
```

Comment:
Ultimately that's for the @NixOS/systemd team to decide, but adding tests like this by abusing VM tests is really asking forl trouble since it also relies on internals (eg. lazy eval of services based on `enabled`). While I'm not sure whether `pkgs/test/systemd.nix` is the best location for that, it's IMHO better than having it here.

Reply:
You're most definitively right. And yes there was a trick here with the underlying assumption that the broken unit will not be used due to `enable=false`.
I glanced over `pkgs/test/` yesterday but did not found enough supporting material to be sure that it could be done properly, due to the need to access the nixos machinery from the pkgs machinery. I took a second look and came up with something cleaner and faster, so thank you for insisting ;-)